File: server/internal/usecase/admin/dto/async_task.go
package dto

import "time"

type AsyncTaskListQuery struct {
	Page       int    `form:"page" validate:"omitempty,min=1"`
	PerPage    int    `form:"per_page" validate:"omitempty,min=1,max=100"`
	TaskType   string `form:"task_type" validate:"omitempty,max=64"`
	Status     string `form:"status" validate:"omitempty,oneof=pending running succeeded failed cancelled"`
	ObjectType string `form:"object_type" validate:"omitempty,max=64"`
	ObjectNo   string `form:"object_no" validate:"omitempty,max=64"`
	DateFrom   string `form:"date_from" validate:"omitempty,max=32"`
	DateTo     string `form:"date_to" validate:"omitempty,max=32"`
}

type AsyncTaskItem struct {
	TaskNo           string     `json:"task_no"`
	TaskType         string     `json:"task_type"`
	Status           string     `json:"status"`
	ObjectType       *string    `json:"object_type"`
	ObjectNo         *string    `json:"object_no"`
	ScheduledAt      time.Time  `json:"scheduled_at"`
	Attempts         int        `json:"attempts"`
	MaxAttempts      int        `json:"max_attempts"`
	LastErrorCode    *string    `json:"last_error_code"`
	LastErrorMessage *string    `json:"last_error_message"`
	LockedBy         *string    `json:"locked_by"`
	LockedUntil      *time.Time `json:"locked_until"`
	CreatedAt        time.Time  `json:"created_at"`
	CompletedAt      *time.Time `json:"completed_at"`
}

type AsyncTaskRetryRequest struct {
	Remark *string `json:"remark" validate:"omitempty,max=500"`
}

File: admin/src/api/async-task.ts
import { http, type ApiEnvelope } from '../utils/request'
import type { PaginatedData } from './admin-user'

export type AsyncTaskStatus = 'pending' | 'running' | 'succeeded' | 'failed' | 'cancelled'

export interface AsyncTaskItem {
  task_no: string
  task_type: string
  status: AsyncTaskStatus
  object_type: string | null
  object_no: string | null
  scheduled_at: string
  attempts: number
  max_attempts: number
  last_error_code: string | null
  last_error_message: string | null
  locked_by: string | null
  locked_until: string | null
  created_at: string
  completed_at: string | null
}

export async function getAsyncTasks(params?: Record<string, unknown>) {
  const response = await http.get<ApiEnvelope<PaginatedData<AsyncTaskItem>>>('/async-tasks', { params })
  return response.data.data
}

export async function retryAsyncTask(taskNo: string, remark?: string) {
  const response = await http.post<ApiEnvelope<AsyncTaskItem>>(`/async-tasks/${taskNo}/retry`, { remark })
  return response.data.data
}
